### ISSUE TITLE
Marshalling strings with certain characters fails when inside a struct

### ISSUE THREAD

**Issue by ducakar** [dotnet/runtime#1355]
Affected versions: .NET Core 3.1
OS: Arch Linux, Buildroot

Marshalling a struct containing a string field via P/Invoke fails when the string contains certain unicode characters. The following exception is thrown:

```
Unhandled exception. System.Runtime.InteropServices.COMException (0x8007007A): The data area passed to a system call is too small.
 (0x8007007A)
   at System.StubHelpers.ValueClassMarshaler.ConvertToNative(IntPtr dst, IntPtr src, IntPtr pMT, CleanupWorkListElement& pCleanupWorkList)
   at Program.foo(Label label)
   at Program.Main(String[] args) in /home/davorin/Razvoj/Concepts/Test/Program.cs:line 7
```

It seems the issue is only triggered when the string contains characters that occupy more than 2 bytes in UTF-16. It doesn't matter whether the struct is passed by value, by pointer or in an array.

Minimal test case:

```cs
using System.Runtime.InteropServices;

class Program
{
    static void Main()
    {
        foo(new Label { Text = "●" });
    }

    [DllImport("foo")]
    static extern void foo(Label label);

    [StructLayout(LayoutKind.Sequential)]
    struct Label
    {
        public string Text;
    }
}
```

```c
// Compile as a shared library:
// gcc -shared -o libfoo.so foo.c

struct Label {
  const char* text;
};

void foo(struct Label label) {
}
```

**Comment by jeffschwMSFT** [comment]
cc @jkoritzinsky

**Comment by jkoritzinsky** [comment]
This is already fixed in master (validated on 3c0cea3b5b329f4ed2e7ed32b9d4de3dc41f6f35). This was likely fixed as part of dotnet/coreclr#26340.
